lib/deimos/active_record/batch_consumption.rb

```diff
@@ -0,0 +1,179 @@
+                        conflict_target: key_cols,
+                        columns: :all
+                      }
+                    }
+                  end
+
+        @klass.import!(messages, options)
+      end
+
+      # Delete any records with a tombstone.
+      # @param records [Array<Array>] List of [key, nil] pairs for a group of
+      # deleted records.
+      def remove_records(records)
+        clause = deleted_query(records)
+
+        clause.delete_all
+      end
+
+      # Create an ActiveRecord relation that matches all of the passed
+      # records. Used for bulk deletion.
+      # @param records [Array<Array>] List of [key, nil] pairs.
+      # @return ActiveRecord::Relation Matching relation.
+      def deleted_query(records)
+        keys = records.
+          map { |k, _| record_key(k)[@klass.primary_key] }.
+          reject(&:nil?)
+
+        @klass.unscoped.where(@klass.primary_key => keys)
+      end
+
+      # Get the set of attribute names that uniquely identify messages in the
+      # batch. Requires at least one record.
+      # @param records [Array<Array>] Non-empty list of [key, value] pairs.
+      # @return [Array<String>] List of attribute names.
+      # @raise If records is empty.
+      def key_columns(records)
+        raise 'Cannot determine key from empty batch' if records.empty?
+
+        first_key, = records.first
+        record_key(first_key).keys
+      end
+
+    private
+
+      # Maximum number of times to retry a block after encountering a deadlock
+      RETRY_COUNT = 2
+
+      # Split the batch into a series of independent slices. Each slice contains
+      # messages that can be processed in any order (i.e. they have distinct
+      # keys). Messages with the same key will be separated into different
+      # slices that maintain the correct order.
+      # E.g. Given messages A1, A2, B1, C1, C2, C3, they will be sliced as:
+      # [[A1, B1, C1], [A2, C2], [C3]]
+      def _slice_batch(messages)
+        # If no keys, just one big slice
+        if self.class.config[:no_keys]
+          return [messages]
+        end
+
+        ops = messages.group_by { |k, _| k }
+
+        # Take the most recent for each key if consumer is set to `compacted`
+        if self.class.config[:compacted]
+          return [ops.values.map(&:last)]
+        end
+
+        # Find maximum depth
+        depth = ops.values.map(&:length).max || 0
+
+        # Generate slices for each depth
+        depth.times.map do |i|
+          ops.values.map { |arr| arr.dig(i) }.compact
+        end
+      end
+
+      # Retry the given block when encountering a deadlock. For any other
+      # exceptions, they are reraised. This is used to handle cases where
+      # the database may be busy but the transaction would succeed if
+      # retried later.
+      def _retry_deadlock(tags=[])
+        count = RETRY_COUNT
+
+        begin
+          yield
+        rescue ::ActiveRecord::Deadlocked => e
```

Comment:
I know at least in older MySQL you'd get a different error than this for an actual deadlock (it would be a `Mysql2::Error`. Can we confirm that this will work across at least MySQL and Postgres?